templates/startup_script.tpl

```diff
@@ -0,0 +1,182 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# First, make sure the `datalab` and `logger` users exist with their
+# home directories setup correctly.
+useradd datalab -u 2000 || useradd datalab
+useradd logger -u 2001 || useradd logger
+
+# In case the instance has started before, the `/home/datalab` directory
+# may already exist, but with the incorrect user ID (since `/etc/passwd`
+# is saved in a tmpfs and changes after restarts). To account for that,
+# we should force the file ownership under `/home/datalab` to match
+# the current UID for the `datalab` user.
+chown -R datalab /home/datalab
+chown -R logger /home/logger
+
+PERSISTENT_DISK_DEV="/dev/disk/by-id/google-${disk_name}"
+MOUNT_DIR="/mnt/disks/datalab-pd"
+MOUNT_CMD="mount -o discard,defaults $${PERSISTENT_DISK_DEV} $${MOUNT_DIR}"
+
+download_docker_image() {
+  # Since /root/.docker is not writable on the default image,
+  # we need to set HOME to be a writable directory. This same
+  # directory is used later on by the datalab.service.
+  export OLD_HOME=$HOME
+  export HOME=/home/datalab
+  echo "Getting Docker credentials"
+  docker-credential-gcr configure-docker
+  echo "Pulling latest image: ${datalab_docker_image}"
+  docker pull ${datalab_docker_image}
+  export HOME=$OLD_HOME
+}
+
+clone_repo() {
+  echo "Creating the datalab directory"
+  mkdir -p $${{MOUNT_DIR}}/content/datalab
+  echo "Cloning the repo datalab-notebooks"
+  docker run --rm -v "$${{MOUNT_DIR}}/content:/content" \
+    --entrypoint "/bin/bash" ${datalab_docker_image} \
+    gcloud source repos clone datalab-notebooks /content/datalab/notebooks
+}
+
+repo_is_populated() {
+  cd $${MOUNT_DIR}/content/datalab/notebooks
+  git show-ref --quiet
+}
+
+populate_repo() {
+  echo "Populating datalab-notebooks repo"
+  docker run --rm -v "$${{MOUNT_DIR}}/content:/content" \
+    --workdir=/content/datalab/notebooks \
+    --entrypoint "/bin/bash"  ${datalab_docker_image} -c "\
+        echo '.ipynb_checkpoints' >> .gitignore; \
+        echo '*.pyc' >> .gitignore; \
+        echo '# Project Notebooks' >> README.md; \
+        git add .gitignore README.md; \
+        git -c user.email=nobody -c user.name=Datalab \
+          commit --message='Set up initial notebook repo.'; \
+        git push origin master; \
+    "
+}
+
+format_disk() {
+  echo "Formatting the persistent disk"
+  mkfs.ext4 -F -E lazy_itable_init=0,lazy_journal_init=0,discard $${PERSISTENT_DISK_DEV}
+  $${MOUNT_CMD}
+  clone_repo
+  if ! repo_is_populated; then
+    populate_repo
+  fi
+}
+
+checked_format_disk() {
+  echo "Checking if the persistent disk needs to be formatted"
+  if [ -z "$(blkid $${PERSISTENT_DISK_DEV})" ]; then
+    format_disk
+  else
+    echo "Disk already formatted, but mounting failed; rebooting..."
```

Comment:
This logic seems like it could result in an infinite loop. Is there a more determinate approach to handling a mount failure?